Describe Compass security (#300)
kyma-incubator/compass: docs/architecture/authentication-and-authorization.md

# Authentication and Authorization

## Introduction
Currently, communication between the Compass and both Runtimes and Applications are not secured. We need to provide some security possibilities.
We want to secure the Compass using ORY's Hydra and OathKeeper. There would be three ways of authentication:
 - OAuth 2.0 
 - Client Certificates (mTLS)
 - JWT token issued by Dex

 There would be three possibilities for securing the connection. To achieve that, first, we need to integrate Hydra and OathKeeper into the Compass. We also need to implement additional supporting components to make our scenarios valid.

## Architecture

> The following diagram represents architecture of the security in Compass:

Reviewer: The responsibility doesn't change, that's why it's not included in the proposal. Gateway component doesn't influence the architecture here, so it shouldn't be a part of the proposal.

![](./assets/security-architecture.svg)

> ### Tenant mapping handler

Reviewer 2: This section is a bit outdated after our latest discussions regarding creating system_auths db table and returning arrays of auths.

Reviewer: Updated

> It is a OathKeeper [hydrator](https://github.com/ory/docs/blob/525608c65694539384b785355d293bc0ad00da27/docs/oathkeeper/pipeline/mutator.md#hydrator) handler responsible for mapping `client_id` to tenant. It will be built into the Director component, as a separate endpoint responsible for retrieving the tenant ID based on database query.

Reviewer 3: is client_id still valid name?

Reviewer: Fixed

> In order to map client_id from Runtime Agent to tenant, we will query `client_id` from Runtime `agentAuth` JSONB field.

Reviewer: I refactored the section, please take a look

> In order to map client_id from Application to tenant, we need to extend Application model with `auth` JSONB field. Then, we can query `client_id` from Application `auth` JSONB field.

Reviewer: It will, as described on the more detailed diagram. I would keep it as it is on the high-level overview diagram.

To unify tenant mapping, `client_id` will be not only generated in OAuth 2.0 flow, but also in Certificates flow. We will write it into certificate. In a result, these two flows we will map `client_id` to tenant.

### GraphQL security

> The Gateway passes request along with JWT token to Compass GraphQL services, such as Director or Connector. The GraphQL components will have authentication middleware and GraphQL [directives](https://graphql.org/learn/queries/#directives) set up for all GraphQL operations (and some specific type fields, if necessary).

Reviewer 3: let assume, that there are 10 applications for a given tenant, how do we want to restrict that given app can update only itself, and cannot updates information for other apps?

Reviewer: I've added another section

In GraphQL services, there will be a HTTP authentication middleware set up, which validates and decodes JWT token. It puts user scopes and tenant in request context 
(`context.Context`).

![](./assets/graphql-security.svg)

When GraphQL operation is processed, an authorization directive is triggered, before actual GraphQL resolver. It will check if the client has required scopes to do the requested operation. To avoid defining permissions statically in GraphQL schema, a YAML file will be loaded with needed requests. In a reality, it will be a ConfigMap injected to Director/Connector. 

> The following example illustrates how we can implement dynamic comparison between required scopes and request scopes for example mutations, queries and type fields:

Reviewer: Sure, but this is a technical detail which I wouldn't include it in the proposal itself.

> The following example illustrates how we can implement dynamic comparison between required scopes and request scopes for example mutations, queries and type fields:

Reviewer: Okay, I mentioned that we can automate it, but for clarity I kept the example

    healthCheckURL: String
    apis(group: String, first: Int = 100, after: PageCursor): APIDefinitionPage! @secureWithScopes(path: "types.Application.apis")
    eventAPIs(group: String, first: Int = 100, after: PageCursor): EventAPIDefinitionPage! @secureWithScopes(path: "types.Application.eventAPIs")
    documents(first: Int = 100, after: PageCursor): DocumentPage! @secureWithScopes(path: "types.Application.documents")
}
```

The `path` parameter would specify path in YAML file (ConfigMap) with required scopes for a given resolver. For example:
```yaml
queries:
    runtimes: "runtime:list"
    runtime: "runtime:get"
mutations:
    createApplication: "application:create"
    updateApplication: "application:update"
    deleteApplication: "application:delete"
types:
    Application:
        webhooks: "webhook:list"
        apis: "api:list"
        eventAPIs: "eventapi:list"
        documents: "document:list"
```
The actual scopes will be defined later.

## Flows

Each flow will be set up on a separate host via different VirtualService, as currently OathKeeper doesn't support certificates and multiple `Bearer` authenticators.

### OAuth 2.0

> There are two ways of creating a `client_id` and `client_secret` pair in the Hydra, using Hydra's [oauth client](https://github.com/kyma-project/kyma/blob/ab3d8878d013f8cc34c3f549dfa2f50f06502f14/docs/security/03-06-oauth2-server.md#register-an-oauth2-client) or [simple POST request](https://github.com/kyma-incubator/examples/tree/master/ory-hydra/scenarios/client-credentials#setup-an-oauth2-client).

Reviewer 3: when this pair will be created? what if the given pair were compromised, can we reuse client_id?

Reviewer: I clarified it after our discussions.

**Obtaining token:**

> 1. Runtime/Application calls Hydra with encoded credentials (`client_id` is the `runtime ID` or `application ID`) and requested scopes.

Reviewer 3: ``client_id` is the `runtime ID` or `application ID`` 
is this still valid?

Reviewer: Nope, updated

> 1. Runtime/Application calls Hydra with encoded credentials (`client_id` is the `runtime ID` or `application ID`) and requested scopes.

Reviewer 3: which oauth flow is described here?

Reviewer: This is client credentials flow: https://auth0.com/docs/flows/concepts/client-credentials

> 1. Runtime/Application calls Hydra with encoded credentials (`client_id` is the `runtime ID` or `application ID`) and requested scopes.

Reviewer 2: Are we still going to use runtime/application ID as `client_id`?

# Authentication and Authorization

## Introduction
Currently, communication between the Compass and both Runtimes and Applications are not secured. We need to provide some security possibilities.
We want to secure the Compass using ORY's Hydra and OathKeeper. There would be three ways of authentication:
 - OAuth 2.0 
 - Client Certificates (mTLS)
 - JWT token issued by Dex

 There would be three possibilities for securing the connection. To achieve that, first, we need to integrate Hydra and OathKeeper into the Compass. We also need to implement additional supporting components to make our scenarios valid.

## Architecture

The following diagram represents the architecture of the security in Compass:

> ![](./assets/security-architecture.svg)

Reviewer 4: yes, it is a high level diagram which says right now that only runtime and application are designed to call compass APIs. Add user and integration system, or just leave one box and name it "everything" which is app, run, user, int_sys.

> ![](./assets/security-architecture.svg)

Reviewer: Oh, okay. Sorry, I wasn't up to date with the latest diagram. Previously, Application and Runtime weren't placed there.

### Tenant Mapping Handler

It is an OathKeeper [hydrator](https://github.com/ory/docs/blob/525608c65694539384b785355d293bc0ad00da27/docs/oathkeeper/pipeline/mutator.md#hydrator) handler responsible for mapping authentication session data to tenant. It could be built into Director itself, as it will also use the same database. It would be implemented as a separate endpoint, such as `/tenant-mapping`.

> To unify the approach for mapping, we could introduce `authorization_id`, widely used by multiple authentication flows.

Reviewer: > maybe even mention here about system_auths table, and not at the end of that chapter

It will be hard to do it at the beginning, because it's a technical detail. Need for `system_auths` table exists because of the `authorization_id`, so it's better to introduce this concept first.

> To unify the approach for mapping, we could introduce `authorization_id`, widely used by multiple authentication flows.

Reviewer: > how we map user to tenant by system_auths from: "JWT token issued by identity service" chapter "point 3".

It is described in the `JWT token issued by identity service` flow.

> To unify the approach for mapping, we could introduce `authorization_id`, widely used by multiple authentication flows.

Reviewer 4: 1) what's about those could/should?
2) fine
3) does it means user flows, uses system_auths or not? In case of user and local/prodcution what is exactly in system_auths.

> To unify the approach for mapping, we could introduce `authorization_id`, widely used by multiple authentication flows.

Reviewer: 3: On local system we use static ConfigMap, and for production system we must have tenant and scopes in token's payload. So, `system_auths` table is not used in JWT token from identity service flow.

The `authorization_id` is equal to:
- `client_id` in OAuth 2.0 authentication flow 
- `username` in Basic authentication flow
- Common Name (CN) in Certificates authentication flow

> While generating one-time token, `client_id`/`client_secret` pair or basic authentication details for Runtime/Application/IntegrationSystem (using proper GraphQL mutation on Director), an entry in `system_auths` table in Director database is created. The `system_auths` table is used for tenant mapping.

Reviewer: I think it would be better to prepare another document regarding IntegrationSystem itself. I don't want to overload this document with too much pieces of information.

For now I will describe Integration System in Terminology doc. Is that OK for you?

> While generating one-time token, `client_id`/`client_secret` pair or basic authentication details for Runtime/Application/IntegrationSystem (using proper GraphQL mutation on Director), an entry in `system_auths` table in Director database is created. The `system_auths` table is used for tenant mapping.

Reviewer 4: - Yes:, add to terminology (with proper description how if differs from App and Runtime, e.g. works with multitenant/multiapps/multiruntimes management and exists in global scope without tenant assignment)
- And no: please add to diagram as well. And link to terminology first occurrence.

> While generating one-time token, `client_id`/`client_secret` pair or basic authentication details for Runtime/Application/IntegrationSystem (using proper GraphQL mutation on Director), an entry in `system_auths` table in Director database is created. The `system_auths` table is used for tenant mapping.

Reviewer: Done

#### `system_auths` table

> The table is used by Director and Tenant Mapping Handler. It would contain the following fields:

Reviewer 4: Please just add scopes to the system_auths table, right? Or you want to add the at runtime and don't store at all.

> The table is used by Director and Tenant Mapping Handler. It would contain the following fields:

Reviewer: Please see updated document. For now scopes for certificates flow will be hardcoded depending on object type (Application / Runtime / Integration System).

**Obtaining token:**

1. Runtime/Application/IntegrationSystem requests `client_id` and `client_credentials` pair from Director by separate GraphQL mutation. Director generates the pair, registers it in Hydra with proper scopes (defined by object type) and writes it in database.
1. Runtime/Application/IntegrationSystem calls Hydra with encoded credentials (`client_id` is the ID of `SystemAuth` entry related to given Runtime/Application/IntegrationSystem) and requested scopes.
1. If the requested scopes are valid, Runtime/Application/IntegrationSystem receives in response an access token, otherwise receives an error.

**Request flow:**

1. Authenticator calls Hydra for introspection of the token.
1. If the token is valid, OathKeeper sends the request to Hydrator. 
1. Hydrator calls Tenant Mapping Handler hosted by `Director` to get `tenant` based on a `client_id` (`client_id` is the ID of `SystemAuth` entry related to given Runtime/Application/IntegrationSystem) .
1. Hydrator passes response to ID_Token mutator which constructs a JWT token with scopes and `tenant` in the payload.

> 1. The request is then forwarded to the desired component (such as `Director` or `Connector`) through the `Gateway` component.

Reviewer 5: Just a small comment: diagram does not include Connector Service in this flow. IMO we should include it after `Gateway` block as well.

 
![Auth](./assets/oauth2-security-diagram.svg)

**Scopes**

In this authentication flow, scopes are read from OAuth 2.0 access token and written directly in output JWT token. Hydra validates if user can request access token with given scopes.

**Proof of concept:** [kyma-incubator/compass#287](https://github.com/kyma-incubator/compass/pull/287)

### JWT token issued by identity service

**Used by:** User

**Obtaining token:**

User logs in to Compass UI 

**Request flow:**

1. Authenticator validates the token using keys provided by identity service. In production environment, tenant **must be** included in token payload. For local development, the `tenant` property is missing from token issued by Dex.
1. If the token is valid, OathKeeper sends the request to Hydrator.
1. Hydrator calls Tenant Mapping Handler hosted by `Director`, which, in production environment, returns **the same** authentication session (as the `tenant` is already in place). For local development, `tenant` is loaded from ConfigMap, where static `user - tenant` mapping is done.
1. Hydrator passes response to ID_Token mutator which constructs a JWT token with scopes and `tenant` in the payload.

> 1. The request is then forwarded to the desired component (such as `Director` or `Connector`) through the `Gateway` component.

Reviewer 5: Same case, no Connector Service present in the diagram.

We want to secure the Compass using ORY's Hydra and Oathkeeper. There would be two ways of authentication:
 - Oauth 2.0 
 - SSL/TLS Certificates.

## Proposed solution
As mentioned, there would be two possibilities for securing the connection. To achieve that, first, we need to integrate Hydra and Oathkeeper into the Compass. Then we will introduce a new component - Mapping Service.
Mapping Service would be a part of Management Plane Services, responsible for extracting from database a `tenant` based on `client_id`.
We would also use a Hydrator and id_token service which are ORY's mutators.

Having that setup, the flow would be as follows: 

1. Runtime/Application calls the Oathkeeper.
2. Oathkeeper calls Hydrator with `client_id` attached to request.
3. Hydrator calls Mapping Service and gets `tenant` in response.
4. Hydrator calls id_token service with `tenant` attached to the request
5. The id_token service constructs a JWT token with proper scopes and `tenant` injected inside it.
6. The token issued by id_token service can be used to authenticate to the Gateway.

> ![Auth](./assets/compass-auth.svg)

Author: done

### Architecture

> #### Mapping service

Author: I don't think Token service is a good name for service responsible for mapping for example `client_id` to `tenant. What about Tenant service?

> #### Mapping service

Reviewer: Sorry, I meant tenant service 😂

# Authentication and Authorization

## Introduction
Currently, communication between the Compass and both runtimes and applications are not secured. We need to provide some security possibilities.
We want to secure the Compass using ORY's Hydra and OathKeeper. There would be two ways of authentication:

>  - Oauth 2.0 

Reviewer 6: ```suggestion
 - OAuth 2.0 
```

>  - SSL/TLS Certificates.

Reviewer 6: ```suggestion
 - Client Certificates (mTLS)
```

## Proposed solution

> As mentioned, there would be two possibilities for securing the connection. To achieve that, first, we need to integrate Hydra and OathKeeper into the Compass. Then we will introduce a new component - Tenant Service.

Reviewer 6: Too many details in one sentence. I would go with something like this:

As mentioned, there would be two possibilities for securing the connection. To achieve that, first, we need to integrate Hydra and OathKeeper into the Compass. We need to implement additional supporting components to make our scenarios valid. On the Director component, we will implement an OathKeeper hydrator handler that will resolve client_id (which can be runtime ID or application ID) to particular tenant ID.

Reviewer 6: Also, link the hydrator to some external document describing its concept.

Tenant Service would be a part of Management Plane Services, responsible for extracting from database a `tenant` based on `client_id`.

> We would also use a Hydrator and ID_token mutator which are ORY's mutators.

Reviewer 6: Rephrase this to describe the ID token mutator use case in our implementation. Remove the Hydrator from here, it should be mentioned in the previous sentence.

It is a service responsible for mapping `client_id` (OAuth 2.0 flow), `runtime ID` or `application ID` (certificates flow) to `tenant`. There will be a simple HTTP endpoint on the Director component, responsible for retrieving it from database. 

### Managing tenants
There will be a neccessary schema extension. For now, Runtimes have field `runtimeAuth` which has `client_id`. We will need the same for Applications.

## Flows

### OAuth 2.0
There are two ways of creating a `client_id` and `client_secret` pair in the Hydra, using Hydra's [oauth client](https://github.com/kyma-project/kyma/blob/ab3d8878d013f8cc34c3f549dfa2f50f06502f14/docs/security/03-06-oauth2-server.md#register-an-oauth2-client) or [simple POST request](https://github.com/kyma-incubator/examples/tree/master/ory-hydra/scenarios/client-credentials#setup-an-oauth2-client).

Receiving token:
1. Runtime/Application calls Hydra with encoded credentials (`client_id` is the `runtime ID`/`application ID`) and requested scopes.
2. If the requested scopes are valid, Runtime/Application receives in response an access token, otherwise receives an error.

Request flow:
1. OathKeeper calls Hydra for introspection of token.

> 2. If the token was valid, OathKeeper sends the request to Hydrator. 

Reviewer 6: ```suggestion
2. If the token is valid, OathKeeper sends the request to Hydrator. 
```

> 3. Hydrator calls Tenant Service to get `tenant` from a `client_id`.

Reviewer 6: ```suggestion
3. Hydrator calls Tenants handler hosted by `Director` to get `tenant` based on a `client_id`.
```

> 4. Hydrator calls ID_Token mutator which constructs a JWT token with scopes and `tenant` in payload.

Reviewer 6: ```suggestion
4. Hydrator calls ID_Token mutator which constructs a JWT token with scopes and `tenant` in the payload.
```

> 5. ID_Token mutator calls the Compass Gateway.

Reviewer 6: ```suggestion
5. The request is then forwarded to the desired component (`Director` or `Connector`) through the `Gateway` component.
```
